### PR TITLE
[12.0] [ADD] connector_onpay: Integration with OnPay.

### LINKED ISSUE #58: [RFC] connector_onpay
_OnPay (https://onpay.com) built the best online payroll service by combining 30 years of experience as a payroll provider with the latest technology._

This module aims to connect Odoo with OnPay to run payroll. As of today, OnPay API does not support sending input data for payroll run. So we decided to generate a csv file for you to export timesheets and expenses out of Odoo and import it (without opening it) into OnPay.

# Configuration

* Go to Employees
* Set the OnPay ID on your employees
* Go to Timesheets > Configuration > Time Types
* Match each of your time type with an OnPay pay type

# Usage

## On Odoo

* Go to Employees
* Switch to the list view
* Select a list of employees
* In the Action menu, select "Export to OnPay"
* In the popup window, select the pay period. It defaults to the previous half-month period.
* Click on Export

## On OnPay

* Go to https://app.onpay.com/app/login
* Go to Your Company > Payroll > Upload Hours
* Select the file downloaded from Odoo
* Click on "View uploaded files"
* Click "Process" for the file we just imported

# Roadmap

* Allow to configure the pay period

# Technical Specifications

## onpay.pay.type

* Add a char field "Name" (name)
* Add a char field "Code" (code)
* Add a boolean "Treat as cash" (treat_as_cash)
* Load the data provided in the sheet "Input - Pay Types" of the mapping

## task.time.type

* Add a many2one field "OnPay Pay Type" (onpay_pay_type_id ) to onpay.pay.type

## onpay.export

* Create the wizard
* Add a date field "From" (date_from)
* Add a date field "To" (date_to)
* Add a button "Export" to call export_onpay() on hr.employee

## ir.action.server

* Add a server action "Export to OnPay" on hr.employee list view to call the wizard onpay.export

## hr.employee

* Add a char field "OnPay ID" (onpay_id)
* Add a method export_onpay() to:
  * Get the timesheet entries for the selected employees and payroll period entered in the wizard 
  * Get the employee approved expenses for the selected employees and payroll period entered in the wizard
  * Group them by data source (timesheets or expenses) and pay type
  * Generate the files using the mapping
  * Provide the generated file to the user

### PR DESCRIPTION
**Pull request by Chandresh-SerpentCS**
Closes #58

### CODE REVIEW

File: connector_onpay/security/security_view.xml [hunk 1]
```diff
@@ -0,0 +1,5 @@
+<odoo>
+	<record id="group_onpay_allow_export" model="res.groups">
+		<field name="name">Allow to export data for OnPay</field>
+	</record>
+</odoo>
```
Comment (max3903) [inline connector_onpay/security/security_view.xml]: @Chandresh-SerpentCS Can you rename this file `res_groups.xml`?

**Comment by max3903** [comment]
@Chandresh-SerpentCS `wizard` --> `wizards`

File: connector_onpay/security/security_view.xml (hunk 1 above)
Comment (Chandresh-SerpentCS) [reply connector_onpay/security/security_view.xml]: Sure.

File: connector_onpay/wizard/onpay_data_export.py [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+            res_id, context = self.export_data_xls()
+            return {
+                'type': 'ir.actions.act_window',
+                'view_type': 'form',
+                'view_mode': 'form',
+                'res_model': 'wiz.xls.content',
+                'target': 'new',
+                'res_id': res_id.id,
+                'context': context,
+            }
```
Comment (max3903) [inline connector_onpay/wizard/onpay_data_export.py]:
```suggestion
        else:
            res_id, context = self.export_data_xls()
        return {
            'type': 'ir.actions.act_window',
            'view_type': 'form',
            'view_mode': 'form',
            'res_model': 'wiz.xls.content',
            'target': 'new',
            'res_id': res_id.id,
            'context': context,
        }
```

File: connector_onpay/wizards/onpay_data_export.py [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+            res_id, context = self.export_data_xls()
+            return {
+                'type': 'ir.actions.act_window',
+                'view_type': 'form',
+                'view_mode': 'form',
+                'res_model': 'wiz.xls.content',
+                'target': 'new',
+                'res_id': res_id.id,
+                'context': context,
+            }
```
Comment (max3903) [inline connector_onpay/wizards/onpay_data_export.py]:
```suggestion
        if self.file_type == 'csv':
            res_id, context = self.export_data_csv()
        else:
            res_id, context = self.export_data_xls()
        return {
            'type': 'ir.actions.act_window',
            'view_type': 'form',
            'view_mode': 'form',
            'res_model': 'wiz.xls.content',
            'target': 'new',
            'res_id': res_id.id,
            'context': context,
        }
```

File: connector_onpay/views/product_template_view.xml [hunk 4]
```diff
@@ -0,0 +1,12 @@
+<odoo>
+    <record id="view_product_template_onpay_inherit" model="ir.ui.view">
+        <field name="name">product.template.onpay</field>
+        <field name="model">product.template</field>
+        <field name="inherit_id" ref="product.product_template_only_form_view"/>
```
Comment (max3903) [inline connector_onpay/views/product_template_view.xml]:
```suggestion
        <field name="inherit_id" ref="hr_expense.product_product_expense_form_view"/>
```


@Chandresh-SerpentCS This view is not used in Expenses > Configuration > Expenses Products

File: connector_onpay/views/product_template_view.xml [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+        <field name="arch" type="xml">
+            <field name="barcode" position="after">
```
Comment (max3903) [inline connector_onpay/views/product_template_view.xml]:
```suggestion
            <field name="property_account_expense_id" position="before">
```

File: connector_onpay/views/product_template_view.xml [hunk 6]
```diff
@@ -0,0 +1,12 @@
+<odoo>
+    <record id="view_product_template_onpay_inherit" model="ir.ui.view">
+        <field name="name">product.template.onpay</field>
+        <field name="model">product.template</field>
```
Comment (max3903) [inline connector_onpay/views/product_template_view.xml]:
```suggestion
    <record id="view_product_product_onpay_inherit" model="ir.ui.view">
        <field name="name">product.product.onpay</field>
        <field name="model">product.product</field>
```

File: connector_onpay/__manifest__.py [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+    "depends": [
+        "hr_timesheet_time_type", "hr_expense",
+    ],
+    "data": [
+        "security/res_groups.xml",
+        "security/ir.model.access.csv",
+        "data/onpay_pay_type.xml",
+        "views/project_time_type_view.xml",
+        "views/hr_employee_view.xml",
+        "views/product_template_view.xml",
```
Comment (max3903) [inline connector_onpay/__manifest__.py]:
```suggestion
        "views/product_product_view.xml",
```

**Review by max3903 — APPROVED** [review]

File: connector_onpay/models/hr_employee.py [hunk 8]
```diff
@@ -0,0 +1,12 @@
+# Copyright (C) 2020 Open Source Integrators
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import models, fields
+
+
+class HrEmployee(models.Model):
+    _inherit = 'hr.employee'
+
+    onpay_id = fields.Char(
```
Comment (dreispt) [inline connector_onpay/models/hr_employee.py]: Suggestion: `_id`  usually means a many2one field, and may lead to confusion when reading code.
I would prefer to name this `onpay_code`.

File: connector_onpay/models/hr_expense.py [hunk 9]
```diff
@@ -0,0 +1,13 @@
+# Copyright (C) 2020 Open Source Integrators
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import models, fields
+
+
+class HrExpense(models.Model):
+    _inherit = 'hr.expense'
+
+    expense_onpay_id = fields.Many2one(
```
Comment (dreispt) [inline connector_onpay/models/hr_expense.py]: Following on my previous comment, here `onpay_id` is an actual many2one, not a char code...

File: connector_onpay/security/ir.model.access.csv [hunk 10]
```diff
@@ -0,0 +1,3 @@
+id,name,model_id:id,group_id:id,perm_read,perm_write,perm_create,perm_unlink
+access_onpay_pay_type,onpay.pay.type,model_onpay_pay_type,base.group_user,1,1,1,1
```
Comment (dreispt) [inline connector_onpay/security/ir.model.access.csv]: Edit access to configuration data should be limited to the HR Manager.

File: connector_onpay/security/ir.model.access.csv [hunk 11]
```diff
@@ -0,0 +1,3 @@
...
+access_onpay_pay_type_group,onpay.pay.type,model_onpay_pay_type,group_onpay_allow_export,1,1,1,1
```
Comment (dreispt) [inline connector_onpay/security/ir.model.access.csv]: I'm quite sure this access line is unnecessary.

File: connector_onpay/wizards/onpay_data_export.py [hunk 12]
```diff
@@ -0,0 +1,236 @@
...
+    def export_data_csv(self):
+        if self._context.get('active_ids'):
+            emp_obj = self.env['hr.employee']
+            onpay_type_obj = self.env['onpay.pay.type']
+
+            # Timesheet data of employees between selected date
+            self._cr.execute("select employee_id,line_onpay_id,\
+                sum(unit_amount) as unit_amount from \
+                account_analytic_line where \
+                date >= %s and date <= %s and employee_id in %s \
```
Comment (dreispt) [inline connector_onpay/wizards/onpay_data_export.py]: Don't we want only lines where `line_onpay_id IS NOT NULL` ?

File: connector_onpay/wizards/onpay_data_export.py [hunk 13]
```diff
@@ -0,0 +1,236 @@
...
+        }
+
+    @api.multi
+    def export_data_csv(self):
+        if self._context.get('active_ids'):
+            emp_obj = self.env['hr.employee']
+            onpay_type_obj = self.env['onpay.pay.type']
+
+            # Timesheet data of employees between selected date
+            self._cr.execute("select employee_id,line_onpay_id,\
```
Comment (dreispt) [inline connector_onpay/wizards/onpay_data_export.py]: If you use triple-quotes for a multi line string, the `\` will not be needed.

File: connector_onpay/wizards/onpay_data_export.py [hunk 14]
```diff
@@ -0,0 +1,236 @@
...
+
+            # Expense data (approved) of employees between selected date
+            self._cr.execute("select employee_id,expense_onpay_id,\
+                sum(total_amount) as \
+                total_amount from \
+                hr_expense where date >= %s and \
+                date <= %s and state='approved' and employee_id in %s \
+                group by employee_id,expense_onpay_id", (
+                self.date_from, self.date_to,
+                tuple(self._context.get('active_ids')),))
```
Comment (dreispt) [inline connector_onpay/wizards/onpay_data_export.py]: I wonder if these don't also post analytic lines, and we end up with duplicate data?

File: connector_onpay/wizards/onpay_data_export.py [hunk 15]
```diff
@@ -0,0 +1,236 @@
...
+                writer = csv.writer(file)
+                writer.writerow(
+                    ["type", "id", "emp_num", "hours",
+                     "rate", "treat_as_cash"])
+
+                # Write the timesheet data in xls sheet by each employee
+                for data in my_data:
+                    emp_id = emp_obj.browse(data.get('employee_id'))
+                    on_pay_id = onpay_type_obj.browse(
+                        data.get('line_onpay_id'))
```
Comment (dreispt) [inline connector_onpay/wizards/onpay_data_export.py]: Browse inside a loop is drives performance .
Expand the SQL to include the additional fields needed.

File: connector_onpay/wizards/onpay_data_export.py [hunk 16]
```diff
@@ -0,0 +1,236 @@
...
+
+            # Timesheet data of employees between selected date
+            self._cr.execute("select employee_id,line_onpay_id,\
+                sum(unit_amount) as unit_amount from \
+                account_analytic_line where \
+                date >= %s and date <= %s and employee_id in %s \
+                group by employee_id,line_onpay_id", (
+                self.date_from, self.date_to,
+                tuple(self._context.get('active_ids')),))
+            my_data = self._cr.dictfetchall()
```
Comment (dreispt) [inline connector_onpay/wizards/onpay_data_export.py]: Code duplication! Please create a shared method to get the data and avoid this.

**Review by dreispt — COMMENTED** [review]
Good as a first iteration, but there are things to fix.

**Comment by dreispt** [comment]
@Chandresh-SerpentCS Reminder for comments to be addressed. Thanks!

File: connector_onpay/security/ir.model.access.csv (hunk 11 above)
Comment (Chandresh-SerpentCS) [reply connector_onpay/security/ir.model.access.csv]: `group_onpay_allow_export` is a custom group introduced in this module. So I added that group. It is mentioned in the design.

**Review by dreispt — APPROVED** [review]

File: connector_onpay/views/project_time_type_view.xml [hunk 17]
```diff
@@ -5,7 +5,7 @@
         <field name="inherit_id" ref="hr_timesheet_time_type.view_project_time_type_form"/>
         <field name="arch" type="xml">
             <field name="code" position="after">
-                <field name="onpay_pay_type_id"/>
+                <field name="onpay_pay_type_id" required="1"/>
```
Comment (dreispt) [inline connector_onpay/views/project_time_type_view.xml]: Revert this please: the field should not be required

File: connector_onpay/models/hr_employee.py [hunk 18]
```diff
@@ -5,8 +5,8 @@
 
 
 class HrEmployee(models.Model):
-    _inherit = 'hr.employee'
+    _inherit = "hr.employee"
 
     onpay_code = fields.Char(
-        string='OnPay ID',
+        string="OnPay ID",
```
Comment (dreispt) [inline connector_onpay/models/hr_employee.py:11]: If you blackify code, always do that in a separate commit, to make clear the actual code changes

File: connector_onpay/wizards/onpay_data_export.py [hunk 19]
```diff
@@ -13,179 +13,206 @@
...
     def employee_timesheet_data(self, date_from, date_to):
         # Timesheet data of employees between selected date
-        self._cr.execute("""select employee_id,line_onpay_id,
-                sum(unit_amount) as unit_amount from
-                account_analytic_line where
+        self._cr.execute(
+            """select aal.employee_id, aal.line_onpay_id,
+                sum(aal.unit_amount) as unit_amount from
+                account_analytic_line as aal
+                LEFT JOIN hr_employee he ON aal.employee_id = he.id
```
Comment (dreispt) [inline connector_onpay/wizards/onpay_data_export.py]: INNER JOIN will be fine here

File: connector_onpay/wizards/onpay_data_export.py [hunk 20]
```diff
@@ -13,179 +13,206 @@
...
+                where he.onpay_code IS NOT Null and
                 date >= %s and date <= %s and employee_id in %s group by
```
Comment (dreispt) [inline connector_onpay/wizards/onpay_data_export.py]: We are missing the filter for time_type onpay_code is not null

**Review by dreispt — CHANGES_REQUESTED** [review]

File: connector_onpay/wizards/onpay_data_export.py (hunk 20 above)
Comment (AmmarOfficewalaSerpentCS) [reply connector_onpay/wizards/onpay_data_export.py]: In a new code, we required a field in a view level

File: connector_onpay/wizards/onpay_data_export.py [hunk 21]
```diff
@@ -45,8 +45,11 @@ def employee_timesheet_data(self, date_from, date_to):
             """select aal.employee_id, aal.line_onpay_id,
                 sum(aal.unit_amount) as unit_amount from
                 account_analytic_line as aal
-                LEFT JOIN hr_employee he ON aal.employee_id = he.id
+                LEFT JOIN hr_employee he ON aal.employee_id = he.id 
+                LEFT JOIN project_time_type ptt ON aal.time_type_id = ptt.id
                 where he.onpay_code IS NOT Null and
+                aal.time_type_id IS NOT Null and
```
Comment (dreispt) [inline connector_onpay/wizards/onpay_data_export.py]: NOT NULL filter is on the time_type onpay code

File: connector_onpay/wizards/onpay_data_export.py (hunk 20 above)
Comment (dreispt) [reply connector_onpay/wizards/onpay_data_export.py]: onpay_code is optional, and we want to filter out empty values

**Comment by dreispt** [comment]
Still needs fixes.
I don't see the fixes for the Expense export SQL.

Why is there a commit with "Merge branch '12.0' into 12.0_connector_onpay"?
if you wally want to get 12.0 newer commits, you should rebase on 12.0 instead of merging.

**Comment by dreispt** [comment]
Note that TravisCI is red
